test/miscellaneous.jl

```diff
@@ -652,4 +656,54 @@
         one_plus_i_avx!(out2)
         @test out1 == out2
     end
+
+    @testset "Mixed CartesianIndex/Int indexing" begin
+        # A demo similar to the exponential filtering demo from https://julialang.org/blog/2016/02/iteration/,
+        # but with no loop-carried dependency.
+        function smoothdim!(s, x, α, Rpre, irng::AbstractUnitRange, Rpost)
+            ifirst, ilast = first(irng), last(irng)
+            ifirst > ilast && return s
+            for Ipost in Rpost
+                # Initialize the first value along the filtered dimension
+                for Ipre in Rpre
+                    s[Ipre, ifirst, Ipost] = x[Ipre, ifirst, Ipost]
+                end
+                # Handle all other entries
+                for i = ifirst+1:ilast
+                    for Ipre in Rpre
+                        s[Ipre, i, Ipost] = α*x[Ipre, i, Ipost] + (1-α)*x[Ipre, i-1, Ipost]
+                    end
+                end
+            end
+            s
+        end
+        function smoothdim_avx!(s, x, α, Rpre, irng::AbstractUnitRange, Rpost)
+            ifirst, ilast = first(irng), last(irng)
+            ifirst > ilast && return s
+            @avx for Ipost in Rpost
+                # Initialize the first value along the filtered dimension
+                for Ipre in Rpre
+                    s[Ipre, ifirst, Ipost] = x[Ipre, ifirst, Ipost]
+                end
```

Comment:
At least for `d=1` in the test-loop at the bottom of this file, the test failure is due to this loop failing to run for all of `Rpost`; it only runs for `first(Rpost)`. (And I do mean `Rpost` rather than `Rpre`.) I am guessing this is a loop-ordering thing? I looked at the expanded code but struggled to figure out which pieces come from what parts of this code; if you get a chance, mind looking at that? For the 3d test below, test failures occur for `d=1` and `d=3`, whereas it passes on `d=2`. Would probably be best to reenable the 5d (or at least, 4d) case before merging, since you need `d=5` to test the case of indexing with `x[::CI{2}, ::Int, ::CI{2}]` (`CI=CartesianIndex`).

Reply:
I think heterogenous `nops` should either only be allowed for `memload` and `memstore` operations, or perhaps also allowed for `compute` iff `Set(Δidxs)` == `Set([1,n])` (i.e., broadcasting some `nops == 1`s to match the others).

I'd lean toward the former now, for simplicity. That means, maintain the old behavior, except for indexing operations.
That means you could make the `@assert` immediate, but conditional on `!accesses_memory(op)`.

I'll take a deeper look at the code in the next few hours and leave more comments.

Reply:
I found a couple problems. I'll push when I've fixed them and leave a few comments.

Reply:
The `tile=(1,1)` is unnecessary. I just added that while experimenting.
But I think `@avx` is being overly aggressive in unrolling this. I'll run benchmarks later.

Reply:
But it's nice that I don't have to think about cache-friendliness when writing out those loops!